src/capi.rs

```diff
@@ -0,0 +1,91 @@
+use std::slice;
+use cose::decoder::decode_signature;
+use cose::SignatureAlgorithm;
+
+unsafe fn from_raw(ptr: *const u8, len: usize) -> Vec<u8> {
+    slice::from_raw_parts(ptr, len).to_vec()
+}
+
+type VerifyCallback = extern "C" fn(*const u8, /* payload */
+                                    usize, /* payload len */
+                                    *const *const u8, /* cert_chain */
+                                    usize, /* # certs */
+                                    *const usize, /* cert lengths in cert_chain */
+                                    *const u8, /* signer cert */
+                                    usize, /* signer cert len */
+                                    *const u8, /* signature bytes */
+                                    usize, /* signature len */
+                                    u8 /* signature algorithm */)
+                                    -> bool;
+
+#[no_mangle]
+pub unsafe extern "C" fn verify_signature_with_cpp(
+    payload: *const u8,
+    payload_len: usize,
+    cose_signature: *const u8,
+    cose_signature_len: usize,
+    verify_callback: VerifyCallback,
+) -> bool {
+    if payload.is_null() || cose_signature.is_null() || payload_len == 0 ||
+        cose_signature_len == 0
+    {
+        return false;
+    }
+
+    // Build Rust variables from C parameters.
+    let payload = from_raw(payload, payload_len);
+    let cose_signature = from_raw(cose_signature, cose_signature_len);
+
+    // Parse the incoming signature.
+    let cose_signatures = decode_signature(cose_signature, &payload);
+    let cose_signatures = match cose_signatures {
+        Ok(signature) => signature,
+        Err(_) => Vec::new(),
+    };
+    if cose_signatures.len() < 1 {
+        return false;
+    }
+
+    let mut result = true;
+    for cose_signature in cose_signatures {
+        let signature_type = cose_signature.signature_type;
+        // ES256 = 0, ES384 = 1, ES521 = 2, PS256 = 3
+        let signature_type = match signature_type {
+            SignatureAlgorithm::ES256 => 0,
```

Comment:
Should we expose some constants in this API so we don't have to use the actual numerical values?

Reply:
I'd love to but I don't think this is possible. We could add a getter function or something. But that's probably more confusing than helping.